internal/controlplane/handlers_authz.go

```diff
@@ -632,23 +627,134 @@ func (s *Server) UpdateRole(ctx context.Context, req *minder.UpdateRoleRequest)
 		return nil, util.UserVisibleError(codes.InvalidArgument, err.Error())
 	}
 
-	// We may be given a human-readable identifier which can vary over time. Resolve
-	// it to an IDP-specific stable identifier so that we can support subject renames.
-	identity := &auth.Identity{
-		Provider:  nil,
-		UserID:    sub,
-		HumanName: sub,
+	// Validate the subject and email - decide if it's about updating an invitation or a role assignment
+	if sub == "" && email != "" {
+		if flags.Bool(ctx, s.featureFlags, flags.UserManagement) {
+			return s.updateInvite(ctx, targetProject, authzRole, email)
+		}
+		return nil, util.UserVisibleError(codes.Unimplemented, "user management is not enabled")
+	} else if sub != "" && email == "" {
+		// If there's a subject, we assume it's a role assignment update
+		return s.updateRole(ctx, targetProject, authzRole, sub)
+	}
+	return nil, util.UserVisibleError(codes.InvalidArgument, "one of subject or email must be specified")
+}
+
+func (s *Server) updateInvite(
+	ctx context.Context,
+	targetProject uuid.UUID,
+	authzRole authz.Role,
+	email string,
+) (*minder.UpdateRoleResponse, error) {
+	var userInvite db.UserInvite
+	// Get the sponsor's user information (current user)
+	currentUser, err := s.store.GetUserBySubject(ctx, auth.GetUserSubjectFromContext(ctx))
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to get user: %s", err)
+	}
+
+	// Get all invitations for this email and project
+	existingInvites, err := s.store.GetInvitationsByEmailAndProject(ctx, db.GetInvitationsByEmailAndProjectParams{
+		Email:   email,
+		Project: targetProject,
+	})
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "error getting invitations: %v", err)
+	}
+
+	// Exit early if there are no or multiple existing invitations for this email and project
+	if len(existingInvites) == 0 {
+		return nil, util.UserVisibleError(codes.NotFound, "no invitations found for this email and project")
+	} else if len(existingInvites) > 1 {
+		return nil, status.Errorf(codes.Internal, "multiple invitations found for this email and project")
 	}
-	if flags.Bool(ctx, s.featureFlags, flags.IDPResolver) {
-		identity, err = s.idClient.Resolve(ctx, sub)
+
+	// Begin a transaction to ensure that the invitation is updated atomically
+	tx, err := s.store.BeginTransaction()
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "error starting transaction: %v", err)
+	}
+	defer s.store.Rollback(tx)
+
+	// At this point, there should be exactly 1 invitation. We should either update its expiration or
+	// discard it and create a new one
```

Comment:
I'm confused about this logic, and when we extend an invite vs re-create it.  I *think* this code does:

- User A invites foo@email.com to be a Viewer
- 8 days pass (expires) and then User A invites foo@email.com as a Viewer again (*Update*)
- 2 days pass and User A updates the invite for foo@email.com to Editor (*Replace*)
- 5 days pass and User B updates the invite for foo@email.com to Editor (*Update*)
- 12 days pass (expires) and User C renews the invite for foo@email.com to Editor (*Update*)

Is that correct?  Why do we preserve the original invite for users B and C, but reset the creation date when A updates the role?

Do we need both a creation and and an update for the invitation?